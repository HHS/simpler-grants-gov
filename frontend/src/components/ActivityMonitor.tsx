"use client";

import { useUser } from "src/services/auth/useUser";

import { useCallback, useEffect, useRef, useState } from "react";

// when a user is logged in, this will track all clicks and key presses and,
// on each action, if the user's auth token is expiring or expired, will refresh the
// token or log the user out accordingly
export function ActivityMonitor() {
  const { user, refreshIfExpiring, refreshIfExpired, logoutLocalUser } =
    useUser();

  const [listening, setListening] = useState(false);
  const handlerRef = useRef<EventListener>(null);

  const refreshTokenIfExpiringOrLogout = useCallback(() => {
    refreshIfExpired()
      .then((expired) => {
        if (expired) {
          logoutLocalUser();
          return;
        }
        return refreshIfExpiring();
      })
      .catch((e) => {
        console.error("Error refreshing token in response to user activity", e);
      });
  }, [refreshIfExpired, refreshIfExpiring, logoutLocalUser]);

  const addHandlers = useCallback(() => {
    if (!listening) {
      document.addEventListener("click", refreshTokenIfExpiringOrLogout);
      document.addEventListener("keydown", refreshTokenIfExpiringOrLogout);
      setListening(true);
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  const removeHandlers = useCallback(() => {
    if (!listening) {
      return;
    }
    document.removeEventListener("click", handlerRef.current as EventListener);
    document.removeEventListener(
      "keydown",
      handlerRef.current as EventListener,
    );
    setListening(false);
    handlerRef.current = null;
  }, [listening]);

  // when logged in status changes, remove or add handlers
  useEffect(() => {
    if (!user || !user.token) {
      removeHandlers();
      return;
    }
    addHandlers();
    return () => {
      removeHandlers();
    };
  }, [user, addHandlers, removeHandlers]);

  // whenever we are not listening for activity, and the handler function has been updated
  // update the ref, so that we have the right reference ready when we want to remove the handlers later.
  // we do not want to pick up any changes made while listening, as those changes will not corerspond to the
  // handler that is actually attached to the document.
  // note that if we try to do this in addHandlers it somehow doesn't work right, given all of the dependencies involved.
  useEffect(() => {
    if (refreshTokenIfExpiringOrLogout && !listening) {
      handlerRef.current = refreshTokenIfExpiringOrLogout;
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  return null;
}
