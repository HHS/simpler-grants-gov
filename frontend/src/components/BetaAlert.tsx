import { useTranslations } from "next-intl";
import { ReactNode } from "react";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BetaAlertProps = {
  containerClasses?: string;
  heading?: string;
  alertMessage?: string | ReactNode;
};

const BetaAlert = ({
  containerClasses,
  heading,
  alertMessage,
}: BetaAlertProps) => {
  const t = useTranslations("BetaAlert");
  const defaultAlertMessage = t.rich("alert", {
    LinkToGrants: (content) => <a href="https://www.grants.gov">{content}</a>,
  });

  return (
    <div data-testid="beta-alert" className={containerClasses}>
      <GridContainer>
        <USWDSAlert
          type="warning"
          headingLevel="h2"
          heading={heading || t("alertTitle")}
          noIcon
        >
          {alertMessage || defaultAlertMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BetaAlert;
