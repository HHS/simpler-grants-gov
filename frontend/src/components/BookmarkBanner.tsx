import { useTranslations } from "next-intl";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BookmarkBannerProps = {
  containerClasses?: string;
  message?: string;
};

const BookmarkBanner = ({ containerClasses, message }: BookmarkBannerProps) => {
  const t = useTranslations("BookmarkBanner");

  const defaultMessage = (
    <>
      {t("message")}
      <br />
      {t.rich("technicalSupportMessage", {
        mailToGrants: (content) => (
          <a href="mailto:simpler@grants.gov">{content}</a>
        ),
      })}
    </>
  );

  return (
    <div data-testid="bookmark-banner" className={containerClasses}>
      <GridContainer>
        <USWDSAlert type="info" headingLevel="h2" heading={t("title")}>
          {message || defaultMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BookmarkBanner;
