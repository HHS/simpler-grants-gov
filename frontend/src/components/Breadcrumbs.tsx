import {
  Breadcrumb,
  BreadcrumbBar,
  BreadcrumbLink,
  GridContainer,
} from "@trussworks/react-uswds";

export type Breadcrumb = {
  title: string;
  path: string;
};

export type BreadcrumbList = Breadcrumb[];

type Props = {
  breadcrumbList: BreadcrumbList;
};

const microdata = {
  ol: {
    itemScope: true,
    itemType: "https://schema.org/BreadcrumbList",
  },
  li: {
    itemScope: true,
    itemProp: "itemListElement",
    itemType: "https://schema.org/ListItem",
  },
  a: {
    itemScope: true,
    itemProp: "item",
    itemType: "https://schema.org/WebPage",
  },
};

const Breadcrumbs = ({ breadcrumbList }: Props) => {
  const breadcrumArray = breadcrumbList.map((breadcrumbInfo, i) => {
    return (
      <Breadcrumb
        key={breadcrumbInfo.title + "-crumb"}
        current={i + 1 === breadcrumbList.length}
        {...microdata.li}
      >
        {i + 1 !== breadcrumbList.length ? (
          <BreadcrumbLink
            href={breadcrumbInfo.path}
            itemID={breadcrumbInfo.path}
            {...microdata.a}
          >
            {}
            <span itemProp="name">{breadcrumbInfo.title}</span>
          </BreadcrumbLink>
        ) : (
          <span itemProp="name">{breadcrumbInfo.title}</span>
        )}
        <meta itemProp="position" content={(i + 1).toString()} />
      </Breadcrumb>
    );
  });

  return (
    <GridContainer data-testid="breadcrumb">
      <BreadcrumbBar listProps={{ ...microdata.ol }}>
        {breadcrumArray}
      </BreadcrumbBar>
    </GridContainer>
  );
};

export default Breadcrumbs;
