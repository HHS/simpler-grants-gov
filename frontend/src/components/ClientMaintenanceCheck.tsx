"use client";

import { useFeatureFlags } from "src/hooks/useFeatureFlags";

import { useEffect } from "react";

/**
 * Checks for the "major" features being offline, otherwise we're not in maintenance, so redir to homepage.
 */
export default function ClientMaintenanceCheck() {
  const {
    checkFeatureFlag, // An instance of FeatureFlagsManager
  } = useFeatureFlags();

  useEffect(() => {
    if (
      !checkFeatureFlag("searchOff") &&
      !checkFeatureFlag("opportunityOff") &&
      checkFeatureFlag("authOn")
    ) {
      // This piece does not seem reliable, maybe due to re-render or rehyrdrate cylces where we redirect away before the actual client live feature flag is set
      // redirect("/", RedirectType.push);
    }
  }, [checkFeatureFlag]);

  return null;
}
