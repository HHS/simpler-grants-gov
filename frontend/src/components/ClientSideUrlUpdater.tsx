"use client";

import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";

import { useRouter } from "next/navigation";
import { useEffect } from "react";

// This should be used as a way to force a url change from a server component
// supports updating:
// - the full url
// - an individual query param
// - the query string
// Note that if you want to zero out the query or param value, you should explicitly pass "" (empty string)
export function ClientSideUrlUpdater({
  param,
  value,
  url,
  query,
}: {
  param?: string;
  value?: string;
  url?: string;
  query?: string;
}) {
  const router = useRouter();
  const { updateQueryParams, searchParams } = useSearchParamUpdater();
  // if query has been passed in, use that, otherwise pass through the current query
  const updatedQuery = query !== undefined ? query : searchParams.get("query");

  useEffect(() => {
    if (url) {
      router.push(url);
    }
  }, [url, router]);
  useEffect(() => {
    if (param && value !== undefined) {
      updateQueryParams(value, param, updatedQuery);
    }
  }, [param, value, updatedQuery, updateQueryParams]);
  return <></>;
}
