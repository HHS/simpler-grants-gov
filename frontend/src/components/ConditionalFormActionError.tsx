import { ErrorMessage } from "@trussworks/react-uswds";

export function ConditionalFormActionError<
  T extends { [key: string]: string[] },
>({ errors, fieldName }: { errors?: T; fieldName: keyof T }) {
  if (!errors) {
    return;
  }
  const error = errors[fieldName];
  if (error) {
    return <ErrorMessage className="maxw-mobile-lg">{error[0]}</ErrorMessage>;
  }
}
