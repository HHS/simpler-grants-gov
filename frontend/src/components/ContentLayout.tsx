import { Grid, GridContainer } from "@trussworks/react-uswds";

type Props = {
  bottomBorder?: "light" | "dark" | "none";
  children: React.ReactNode;
  className?: string;
  gridGap?: true | "sm" | "md" | "lg" | "2px" | "05" | 1 | 2 | 3 | 4 | 5 | 6;
  paddingTop?: boolean;
  title?: string | React.ReactNode;
  titleSize?: "l" | "m";
};

const ContentLayout = ({
  bottomBorder = "none",
  className = "",
  children,
  gridGap = true,
  paddingTop = true,
  title,
  titleSize = "l",
}: Props) => {
  const formattedTitle = () => {
    if (!title) return "";
    const size =
      titleSize === "l"
        ? "tablet-lg:font-sans-xl desktop-lg:font-sans-2xl"
        : "tablet-lg:font-sans-l desktop-lg:font-sans-xl";
    return (
      <h2
        className={`margin-bottom-0 ${size} ${
          !paddingTop ? "margin-top-0" : ""
        }`}
      >
        {title}
      </h2>
    );
  };

  const bborder =
    bottomBorder === "light"
      ? "border-bottom-2px border-base-lighter"
      : bottomBorder === "dark"
        ? "border-bottom-2px border-base-light"
        : "";

  return (
    <GridContainer
      className={`padding-y-1 tablet:padding-y-3 desktop-lg:padding-y-6 ${
        !paddingTop
          ? "padding-top-0 tablet:padding-top-0 desktop-lg:padding-top-0"
          : ""
      } ${bborder} ${className}`}
    >
      {formattedTitle()}
      <Grid row gap={gridGap}>
        {children}
      </Grid>
    </GridContainer>
  );
};

export default ContentLayout;
