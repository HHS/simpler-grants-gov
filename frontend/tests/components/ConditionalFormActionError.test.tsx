import { render, screen } from "@testing-library/react";

import { ConditionalFormActionError } from "src/components/ConditionalFormActionError";

const TypedConditionalFormActionError = ConditionalFormActionError<{
  someFieldName?: string[];
  anotherFieldName?: string[];
}>;

describe("ConditionalFormActionError", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <TypedConditionalFormActionError
        fieldName="someFieldName"
        errors={{ someFieldName: ["special error"] }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it("does not render error if no errors", () => {
    render(<TypedConditionalFormActionError fieldName="someFieldName" />);
    expect(screen.queryByRole("alert")).not.toBeInTheDocument();
  });
  it("does not render error if no error matches field name", () => {
    render(
      <TypedConditionalFormActionError
        fieldName="someFieldName"
        errors={{ anotherFieldName: ["error"] }}
      />,
    );
    expect(screen.queryByRole("alert")).not.toBeInTheDocument();
  });
  it("renders error for field name", () => {
    render(
      <TypedConditionalFormActionError
        fieldName="someFieldName"
        errors={{ someFieldName: ["special error"] }}
      />,
    );
    expect(screen.getByRole("alert")).toBeInTheDocument();
    expect(screen.getByText("special error")).toBeInTheDocument();
  });
});
