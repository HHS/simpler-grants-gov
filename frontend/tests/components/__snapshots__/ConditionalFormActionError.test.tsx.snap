// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionalFormActionError matches snapshot 1`] = `
<div>
  <span
    class="usa-error-message maxw-mobile-lg"
    data-testid="errorMessage"
    role="alert"
  >
    special error
  </span>
</div>
`;
